include/aos/common/thread.hpp

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef AOS_THREAD_HPP_
+#define AOS_THREAD_HPP_
+
+#include <pthread.h>
+
+#include "aos/common/error.hpp"
+
+namespace aos {
+
+/**
+ * Aos thread.
+ */
+class Thread {
+public:
+    /**
+     * Constructs Aos thread instance.
+     *
+     * @param routine function to be called in thread.
+     * @param arg optional argument that is passed to the thread function.
+     */
+
+    template <typename T>
+    explicit Thread(T routine, void* arg = nullptr)
+        : mArg(arg)
+    {
+        static auto sRoutine = routine;
+
+        mAdapter = [](void* arg) -> void* {
+            sRoutine(arg);
+
+            return nullptr;
+        };
+    }
+
+    /**
+     * Runs thread function.
+     *
+     * @return Error.
+     */
+    Error Run() { return pthread_create(&mPThread, nullptr, mAdapter, mArg); }
+
+    /**
+     * Waits thread function is finished.
+     *
+     * @return Error.
+     */
+    Error Join() { return pthread_join(mPThread, nullptr); }
+
+private:
+    void* (*mAdapter)(void*);
+    void*     mArg;
+    pthread_t mPThread;
+
+    Thread(Thread const&) = delete;
+    Thread& operator=(Thread const&) = delete;
+};
+
+/**
+ * Aos mutex.
+ */
+class Mutex {
+public:
+    /**
+     * Constructs Aos mutex.
+     */
+    Mutex() { pthread_mutex_init(&mPMutex, nullptr); }
+
+    /**
+     * Destroys Aos mutex.
+     */
+    ~Mutex() { pthread_mutex_destroy(&mPMutex); }
+
+    /**
+     * Locks Aos mutex.
+     *
+     * @return Error.
+     */
+    Error Lock() { return pthread_mutex_lock(&mPMutex); }
+
+    /**
+     * Unlocks Aos mutex.
+     *
+     * @return Error.
+     */
+    Error Unlock() { return pthread_mutex_unlock(&mPMutex); }
+
+private:
+    friend class ConditionalVariable;
```

Comment:
Additionally, you could define a conversion operator from Mutex to pthread_mutex_t, which would allow you to use the Mutex object directly inside the ConditionalVariable class without needing to declare it as a friend.

Reply:
ok

Reply:
done